email_notifications: Fix missed message email notifications of the welcome bot.

prakhar1144 opened the pull request:
A missed message email notification, where the message is the welcome message sent by the welcome bot on account creation, get sent when the user somehow not focuses the browser tab during account creation.

No missed message email notification should be sent for the messages generated by the welcome bot.

A separate function named `trivially_should_not_notify` is added which extracts the common checks from `get_push_notification_trigger` and `get_email_notification_trigger` which are users' notification settings
independent and thus don't depend on what type of notification (email/push) it is.

A check is introduced in `trivially_should_not_notify`, not to notify if `sender_is_welcome_bot` is true.

TestCases are updated to include the `sender_is_welcome_bot` check in the early (False) return patterns of `is_push_notifiable` and `is_email_noitifiable`.

**One query reduced** for both `test_create_user_with_multiple_streams` and `test_register`.

**Reason:** When welcome bot sends message after user creation `do_send_messages` calls [`get_active_presence_idle_user_ids`](https://github.com/zulip/zulip/blob/aebdf6af8c6675fbd2792888d701d582c4a1110a/zerver/actions/message_send.py#L732), [`user_ids`](https://github.com/zulip/zulip/blob/aebdf6af8c6675fbd2792888d701d582c4a1110a/zerver/actions/message_send.py#L758) in `get_active_presence_idle_user_ids` remains empty if
`sender_is_welcome_bot` is true because `is_notifiable` returns false. `get_active_presence_idle_user_ids` calls `filter_presence_idle_user_ids` and since the `user_ids` is empty, the query inside the function doesn't
get executed. 

**Manual Testing:** The issue is hard to reproduce in a UI. To reproduce the issue, raise an exception in zerver/views/home.py by adding raise Exception("Throw exception after account creation") before return web_public_view(home_real)(request) [reference](https://github.com/zulip/zulip/blob/2ff4143aa4cd68a1eb338455ef005f9c6bec17a8/zerver/views/home.py#L131)

Create a new account, then visit \emails to see the missed message email notification. No more missed message email notification get received after including the changes.

Fixes: #22884

<hr>

**Screenshots and screen captures:**

**Before**
- <details>
  <summary>Terminal: missed_message email sent.</summary>
   <img src="https://user-images.githubusercontent.com/56781761/197336688-84cd6876-f04b-4ba1-a076-272d48d457be.png"> 
   </img>
  </details>

- <details>
  <summary>Email</summary>
   <img src="https://user-images.githubusercontent.com/56781761/197336710-a3ed1fbf-1e74-4da6-a5af-9872c74b9b3c.png"> 
   </img>
  </details>

**After**
- <details>
  <summary>Terminal: No email sent.</summary>
   <img src="https://user-images.githubusercontent.com/56781761/197336727-4b292aee-4801-4c74-b6e5-a0fc80904723.png"> 
   </img>
  </details>

<hr>

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

prakhar1144 commented:
You can have a look at the **backend test failing** section of the above comment.
I am still trying to find out the potential reason for the failing tests. 
**Any help is appreciated.**

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -970,7 +972,9 @@ def get_client_payload(apply_markdown: bool, client_gravatar: bool) -> Dict[str,
-        if not user_notifications_data.is_notifiable(acting_user_id=sender_id, idle=True):
+        if is_sender_welcome_bot or not user_notifications_data.is_notifiable(
+            acting_user_id=sender_id, idle=True
+        ):
```
We likely need a similar check in `missedmessage_hook`... or perhaps it'd be better to make `maybe_enqueue_notifications` accept this parameter and do this check.

prakhar1144 replied:
As per [docs](https://zulip.readthedocs.io/en/latest/subsystems/notifications.html#the-mobile-email-notifications-flow):
> We try to contain the “when to notify” business logic in the zerver/lib/notification_data.py class methods.

So, ideally the logic should reside in the [`is_email_notifiable()`](https://github.com/zulip/zulip/blob/cebf43903c5fc676432a600281fb18a7e192f3af/zerver/lib/notification_data.py#L136) function.

Possible solutions:

1. In `maybe_enqueue_notification`, update [this logic ](https://github.com/zulip/zulip/blob/cebf43903c5fc676432a600281fb18a7e192f3af/zerver/tornado/event_queue.py#L830)as:

> if **not is_sender_welcome_bot and** user_notifications_data.is_email_notifiable(acting_user_id, idle):

2. pass `is_sender_welcome_bot` as the third argument to `is_email_notifiable` and perform the check inside the function. (update test cases accordingly)

3. fetch `welcome_bot` id inside the `is_email_notifiable` and perform the check. But this will lead to an added query. So, this should be avoided.

I think 2nd approach is the best. Do you see any shortcomings?

timabbott replied:
@abhijeetbodas2001 @PIG208 @akashaviator I'd be curious for your thoughts on where this should sit.

akashaviator replied:
IMO, `is_email_notifiable()` should check this. It should always be able to return the result it is supposed to give.

timabbott replied:
Yeah I think I agree with that reasoning. @prakhar1144 can you try reworking this to that model?

PIG208 replied:
I think we can extract a separate method `should_notify` or something, from generic checks that both `is_push_notifiable` and `is_email_notifiable` [do](https://github.com/zulip/zulip/blob/main/zerver/lib/notification_data.py#L113-L120)), which includes the welcome bot check we are planning for right now.

prakhar1144 replied:
After considering everyone's suggestions. This is what I'm planning:

* A separate method named `should_notify` or `should_sender_notify` extracts the generic checks i.e 
```
  if self.user_id == acting_user_id:
      return None

  if self.sender_is_muted:
      return None

  if is_sender_welcome_bot:
      return None
```
* Pass `is_sender_welcome_bot` as the last argument to `is_notifiable` and `maybe_enqueue_notifications`.

* In the case of `process_message_event`, the check will stop [here itself](https://github.com/zulip/zulip/blob/8883b643881f0f243712ad020d22ebd400fb0b94/zerver/tornado/event_queue.py#L973) if `sender_is_welcome_bot` thus avoiding the [expensive operation](https://github.com/zulip/zulip/blob/8883b643881f0f243712ad020d22ebd400fb0b94/zerver/tornado/event_queue.py#L969)

* In the case of [`missedmessage_hook`](https://github.com/zulip/zulip/blob/8883b643881f0f243712ad020d22ebd400fb0b94/zerver/tornado/event_queue.py#L716), `maybe_enqueue_notifications` will take care of the checks.

Let me know if it sounds good to you guys.

prakhar1144 replied:
@timabbott @PIG208 

Please leave feedback so I can proceed. It would be greatly appreciated.

abhijeetbodas2001 replied:
> I think we can extract generic checks that both `is_push_notifiable` and `is_email_notifiable` [have](https://github.com/zulip/zulip/blob/main/zerver/lib/notification_data.py#L116-L120), and create a separate method called `should_notify` or something. This helper should then include the welcome bot check we are planning for right now.

That makes sense to me, except that we could use something more descriptive than `should_notify` (like `trivially_should_not_notify` or something like that, since it will contain trivial checks like muted user/welcome bot which don't involve any notification settings).
@prakhar1144 you can probably add a prep commit to pull out the already existing common checks into a separate function and then add the welcome bot check to that function.

abhijeetbodas2001 replied:
One question I have is -- why not make `is_sender_welcome_bot` a property of `UserMessageNotificationsData` itself? For context, one of the motivations behind making `UserMessageNotificationsData` was that having so many parameters for notification checks had become unwieldy in `event_queue.py` :)

PIG208 replied:
This sounds right to me.

prakhar1144 replied:
> One question I have is -- why not make `is_sender_welcome_bot` a property of `UserMessageNotificationsData` itself? For context, one of the motivations behind introducing `UserMessageNotificationsData` was that having so many parameters for notification checks and passing them around functions in `event_queue.py` was becoming unwieldy :)

Hm, Makes Sense!

prakhar1144 commented:
@abhijeetbodas2001 I have updated the PR. Can you please review this? Thanks.

abhijeetbodas2001 reviewed (commented):
Thanks for updating! The code itself looks good to me. Some minor comments:
* I didn't understand why the number of queries reduces in the tests. Could you elaborate on that in the commit message?
* It would be better to add the new function and extract the common checks in a separate, prep commit before adding the `is_sender_welcome_bot` check
* It might be helpful to add a docstring to the new function saying what "trivially" here really is (checks which don't involve any of the user's notification settings and thus don't depend on what type of notification (email/mobile) it is)
* You could omit the details of the code flow from the commit message -- those are clear from reading the code itself
* One minor this is, `sender_is_welcome_bot` might be better than `is_sender_welcome_bot` as a variable name, but that's probably nitpicking, so feel free to ignore this point if you wish

prakhar1144 commented:
@abhijeetbodas2001 I have updated the PR as per your suggestions. Please have a look and let me know for any changes.

abhijeetbodas2001 commented:
Thanks! This looks good to me 🙂

alya commented:
@timabbott , @abhijeetbodas2001 's feedback has been addressed, so I think this PR is ready for another look from you.

prakhar1144 commented:
Hi @timabbott, awaiting your another review on this. Thanks.

timabbott commented:
This looks pretty good, but it's enough of a feature that I think we should leave it for post 6.0. Added a label so we merge it then.

prakhar1144 commented:
@timabbott 
> This looks pretty good, but it's enough of a feature that I think we should leave it for post 6.0. Added a label so we merge it then.

I think this should be good to merge now.

timabbott commented:
Thinking about the semantics, if we merge this, then ALL messages by Welcome Bot will not be eligible for email/push notifications. I think that's correct (we intend Notification Bot to be used for non-onboarding features), but I'd like to figure out where to record that fact such that we don't in the future decide to use Welcome Bot for something where that'd be a problem.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -887,12 +887,14 @@ class UserData(TypedDict):
                 user_id=user_id,
                 flags=user_flags.get(user_id, []),
                 private_message=(message_type == "private"),
+                sender_is_welcome_bot=sender_is_welcome_bot,
```
Why did you put this here in the sequence of parameters? I think at the end might be better for readability.

prakhar1144 replied:
> Why did you put this here in the sequence of parameters?

* To keep boolean fields (private_message, sender_is_welcome_bot) at one place.
* Set[int] (*_user_ids) fields at one place.

Hence, Keeping it consistent with [this](https://github.com/zulip/zulip/pull/23225/files#diff-8319130d25323802a8a47fa49f43d3c5e4c73567aa1e2fc95adb6d4afa68646aL24)

timabbott replied:
OK, seems reasonable I guess.

timabbott commented:
I added a comment and pushed back here. But thinking about this a bit more -- I think we might want to do this a bit differently. 

Rather than hardcoding Welcome Bot as being special, I think a cleaner specification would be for `internal_send_private_message` to accept a parameter `disable_external_notifications=True`, rather than this extremely specific Welcome Bot detail -- that way, we can potentially control this behavior for specific call points in the future.

Can you rework this pull request with that model? Thanks!

prakhar1144 commented:
@abhijeetbodas2001 
We have a subtle change in the model of how we're handling this.

As `Tim Abbott` said in the previous comment:
> Rather than hardcoding Welcome Bot as being special, I think a cleaner specification would be for internal_send_private_message to accept a parameter disable_external_notifications=True, rather than this extremely specific Welcome Bot detail -- that way, we can potentially control this behavior for specific call points in the future.

I have updated the PR as per [sugggestion](https://github.com/zulip/zulip/pull/23225#issuecomment-1399050136).

Can you please review this PR once. -- Thanks!

abhijeetbodas2001 commented on zerver/lib/onboarding.py:
```diff
@@ -92,8 +92,16 @@ def send_initial_pms(user: UserProfile) -> None:
 
+    # Note: Welcome bot doesn't trigger email/push notifications,
+    # as this is intended to be seen contextually in the application.
+    # Set 'disable_external_notifications = True' for the implementation of
+    # this rule.
```
You might want to remove this part of the comment -- the code itself is pretty clear :)
(Same for the comment in `send_welcome_bot_response`)

abhijeetbodas2001 commented on zerver/tests/test_tutorial.py:
```diff
@@ -15,7 +15,13 @@ def setUp(self) -> None:
+            welcome_bot,
+            user,
+            content,
+            disable_external_notifications=disable_external_notifications,
```
Might be more concise to just send `True` as positional argument?

abhijeetbodas2001 commented on zerver/tornado/event_queue.py:
```diff
@@ -1128,6 +1133,7 @@ def process_message_update_event(
     wildcard_mention_user_ids = set(event_template.pop("wildcard_mention_user_ids", []))
     muted_sender_user_ids = set(event_template.pop("muted_sender_user_ids", []))
     all_bot_user_ids = set(event_template.pop("all_bot_user_ids", []))
+    disable_external_notifications = event_template.get("disable_external_notifications", False)
```
Like other fields, this should probably be `event_template.pop` instead of `get` so that we don't send this unnecessary data to the client as event data later on.

abhijeetbodas2001 reviewed (commented):
Thanks for updating the PR. This indeed seems like a better approach. I've reviewed the PR and left some minor comments, but looks good otherwise.
One more thing that we should do is add a new end-to-end test for this scenario (message from Welcome bot) to `zerver/tests/test_event_queue.py/MissedMessageHookTest` so that we know that the value of this new field is passed correctly up to the event queue code -- could you please do that?

prakhar1144 commented:
@abhijeetbodas2001 Thanks for the review.

I have updated the PR with your suggested minor updates and Test case addition.

**I am facing a problem, while implementing the [test case](https://github.com/zulip/zulip/pull/23225/files#diff-b7ecab40a23cd87e0acf58e7c8220330787bb4f7bb59e8ba601ae648ffdc968aR679).**

* The new test case I have implemented calls `internal_send_private_message` which returns `Optional[int]`.
* The already present test cases in `MissedMessageHook` calls `self.send_personal_message` or `self.send_stream_message` which returns `int`

Now the [message_id](https://github.com/zulip/zulip/pull/23225/files#diff-b7ecab40a23cd87e0acf58e7c8220330787bb4f7bb59e8ba601ae648ffdc968aL160) in `assert_maybe_enqueue_notifications_call_args` needs to be `int`.

But for time being I have updated it to `Any` to bypass linter error.

Can you suggest some workaround? -- Thanks.

abhijeetbodas2001 commented:
You can put a `assert message_id is not None` and mypy will understand that you expect it to definitely be an `int` (and the static checks will pass).
More mypy tips at https://zulip.readthedocs.io/en/latest/testing/mypy.html!

prakhar1144 commented:
@abhijeetbodas2001 It works -- Thanks!

prakhar1144 commented:
@timabbott 
I have updated the PR as per your suggestions.

`abhijeetbodas2001` has done another round of review.

You can have a look on this. -- Thanks.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1603,13 +1612,22 @@ def internal_prep_private_message(
+    sender: UserProfile,
+    recipient_user: UserProfile,
+    content: str,
+    disable_external_notifications: bool = False,
```
We should use a "mandatory keyword argument" here -- i.e. `content: str, *, disable_external_notifications: bool = False`, and callers that want to set this parameter should pass it explicitly as `disable_external_notificaitons=True`. This pattern makes for MUCH more readable code, since you can see what this `True` does without going and reading the function it's calling. 

The same applies for all the other `internal_send_*|internal_prep_*` functions -- all the parameters after `content` should have this rule. It might make sense to do a preparatory refactor commit to make the existing parameters for the other functions that have them follow this pattern, to keep the new code simple.

timabbott commented on zerver/lib/onboarding.py:
```diff
@@ -214,7 +219,15 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
-    internal_send_private_message(welcome_bot, send_request.message.sender, content)
+
+    # Note: Welcome bot doesn't trigger email/push notifications,
+    # as this is intended to be seen contextually in the application.
```
Let's move this comment inside the call, just before the new parameter -- that comment placement is more readable.

timabbott commented on zerver/tests/test_event_queue.py:
```diff
@@ -674,6 +675,28 @@ def test_bot_recipient(self) -> None:
+    def test_disable_external_notifications(self) -> None:
+        # Email and push notifications should not be sent.
+        # Currently, external notifications are disabled for PMs sent by welcome bot.
+        # In general, it should be true for any sender, with the parameter set to true.
```
This comment is confusing -- welcome bot is no longer relevant to what we're testing here.

# The disable_external_notifications parameter, used for messages sent by welcome bot, should
# result in no email/push notifications being sent regardless of the message type.

timabbott commented on zerver/tests/test_event_queue.py:
```diff
@@ -674,6 +675,28 @@ def test_bot_recipient(self) -> None:
+                pm_email_notify=True,
+                pm_push_notify=True,
+                disable_external_notifications=True,
+                already_notified={"email_notified": False, "push_notified": False},
```
Does this part of the test have a different result if external notifications were not enabled? It's worth calling attention in a comment to the fact that (I think) this last `already_notified` line is False because of this parameter, since it's otherwise kinda unclear what exactly we're testing...

abhijeetbodas2001 replied:
> external notifications were not enabled?

You mean not disabled? Right, if they weren't disabled, the `already_notified` values would both be true. If we're going to add a comment explaining this, would it be better to add it as a docstring to this test class (since the other tests here also follow this pattern of asserting the `UserNessageNotificatioData` instance fields and also asserting the `already_notified` values)?

timabbott replied:
Yeah, that's what I meant. Maybe we're good here even without a new comment, just wanted to make sure I was reading this right.

timabbott commented on zerver/tests/test_notification_data.py:
```diff
@@ -113,6 +113,23 @@ def test_is_push_notifiable(self) -> None:
         )
         self.assertFalse(user_data.is_push_notifiable(acting_user_id=acting_user_id, idle=True))
 
+        # Message sender is the welcome bot.
```
This comment needs updating -- it should explain something like "disable_external_notifications takes precedence over  other flags."

timabbott commented on zerver/tests/test_tutorial.py:
```diff
@@ -15,7 +15,12 @@ def setUp(self) -> None:
+            welcome_bot,
+            user,
+            content,
+            True,
```
In addition to making this call via keyword argument ... we should explain why this parameter is set, and why it's important (will the tests fail if we don't have it here, or will it just be a bad simulation?).

prakhar1144 commented:
@timabbott 

I have made the suggested changes.

ps: Awaiting your response on **abhijeetbodas** [comment](https://github.com/zulip/zulip/pull/23225#discussion_r1084770766)

timabbott commented:
This is great, merged with a slight change to one of the comments, thanks @prakhar1144! Thanks for the careful iteration in what is a fairly complex code path.

You might want to chat with @abhijeetbodas2001 in PMs about good next projects in this code path - stuff like "Follow topic", and then start a thread in #backend about what you'd like to work on next.

(I force-pushed to 1a400b21e72ca50578d3a45f30be79b258e9cb2c since I'd forgotten to apply a small docstring clarification tweak)
